patches/minecraft/net/minecraft/item/SpawnEggItem.java.patch

```diff
@@ -0,0 +1,82 @@
++      super(builder);
++      this.field_200890_d = typeIn;
++      this.field_195988_c = primaryColorIn;
++      this.field_195989_d = secondaryColorIn;
++      MOD_EGGS.add(this);
++   }
++
+    public ActionResultType func_195939_a(ItemUseContext p_195939_1_) {
+       World world = p_195939_1_.func_195991_k();
+       if (!(world instanceof ServerWorld)) {
+@@ -131,18 +_,18 @@
+    }
+ 
+    public static Iterable<SpawnEggItem> func_195985_g() {
+-      return Iterables.unmodifiableIterable(field_195987_b.values());
++      return Iterables.unmodifiableIterable(Iterables.concat(field_195987_b.values(), MOD_EGGS));
+    }
+ 
+    public EntityType<?> func_208076_b(@Nullable CompoundNBT p_208076_1_) {
+       if (p_208076_1_ != null && p_208076_1_.func_150297_b("EntityTag", 10)) {
+          CompoundNBT compoundnbt = p_208076_1_.func_74775_l("EntityTag");
+          if (compoundnbt.func_150297_b("id", 8)) {
+-            return EntityType.func_220327_a(compoundnbt.func_74779_i("id")).orElse(this.field_200890_d);
++            return EntityType.func_220327_a(compoundnbt.func_74779_i("id")).orElse(this.field_200890_d.get());
+          }
+       }
+ 
+-      return this.field_200890_d;
++      return this.field_200890_d.get();
+    }
+ 
+    public Optional<MobEntity> func_234809_a_(PlayerEntity p_234809_1_, MobEntity p_234809_2_, EntityType<? extends MobEntity> p_234809_3_, ServerWorld p_234809_4_, Vector3d p_234809_5_, ItemStack p_234809_6_) {
+@@ -178,4 +_,26 @@
+          }
+       }
+    }
++
++   /* ======================================== FORGE START ===================================== */
++   public static void finalizeModEggs() {
++      net.minecraft.dispenser.DefaultDispenseItemBehavior eggDispenseBehaviour = new net.minecraft.dispenser.DefaultDispenseItemBehavior() {
++         public ItemStack func_82487_b(net.minecraft.dispenser.IBlockSource source, ItemStack stack) {
++            Direction facing = source.func_189992_e().func_177229_b(net.minecraft.block.DispenserBlock.field_176441_a);
++            EntityType<?> entitytype = ((SpawnEggItem)stack.func_77973_b()).func_208076_b(stack.func_77978_p());
++            if (entitytype == null) { return stack; } //TODO - 1.17: Remove check for null and enforce it being nonnull
++
++            entitytype.func_220331_a(source.func_197524_h(), stack, null, source.func_180699_d().func_177972_a(facing), SpawnReason.DISPENSER, facing != Direction.UP, false);
++            stack.func_190918_g(1);
++            return stack;
++         }
++      };
++
++      MOD_EGGS.forEach(egg -> {
++         field_195987_b.put(egg.func_208076_b(null), egg);
++         net.minecraft.block.DispenserBlock.func_199774_a(egg, eggDispenseBehaviour);
++      });
++      MOD_EGGS.clear();
++   }
++   /* ========================================= FORGE END ====================================== */
```

Comment:
Couldn't all this (including `MOD_EGGS`) be in say `ForgeHooks` and make the whole patch a lot smaller? Additionally, how would a mod override the default dispense behavior you are adding here?

Reply:
The code in this method was originally in the registry bake callback and @pupnewfster mentioned that neither `MOD_EGGS` nor the original spawn eggs map should be public API which is why I moved it into `SpawnEggItem` itself.
If everything worked as you would expect from the implementation of the vanilla spawn eggs, a mod would have absolutely no reason to register a dispense behaviour for their spawn eggs. Apart from that, if I haven't missed anything, a mod registering a dispense behaviour in `FMLCommonSetupEvent` would overwrite this anyway due to the fact that the dispense behaviour is registered in a bake callback and the dispense behaviour "registry" is just a item->behaviour map.

Reply:
Registry bake will occur multiple times, for example it will run whenever you load a world. So no, you cannot override it with `FMLCommonSetupEvent`.

Reply:
Also, you can make fields private in any class, so just because you move `MOD_EGGS` to e.g. `ForgeHooks` doesn't mean it suddenly can no longer be private.

Reply:
That is true in this case for `MOD_EGGS` but not for the `BY_ID` map.

Reply:
I decided to keep this code in the patch for simplicity and encapsulation.